managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -848,6 +874,97 @@ private long consumedLedgerSize(long ledgerSize, long ledgerEntries, long consum
     }
 
     @Override
+    public synchronized void asyncTerminate(TerminateCallback callback, Object ctx) {
+        if (state == State.Fenced) {
+            callback.terminateFailed(new ManagedLedgerFencedException(), ctx);
+            return;
+        } else if (state == State.Terminated) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Ignoring request to terminate an already terminated managed ledger", name);
+            }
+            callback.terminateComplete(lastConfirmedEntry, ctx);
+            return;
+        }
+
+        log.info("[{}] Terminating managed ledger", name);
+        state = State.Terminated;
+
+        LedgerHandle lh = currentLedger;
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Closing current writing ledger {}", name, lh.getId());
+        }
+
+        mbean.startDataLedgerCloseOp();
+        lh.asyncClose((rc, lh1, ctx1) -> {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Close complete for ledger {}: rc = {}", name, lh.getId(), rc);
+            }
+            mbean.endDataLedgerCloseOp();
+            if (rc != BKException.Code.OK) {
+                callback.terminateFailed(new ManagedLedgerException(BKException.getMessage(rc)), ctx);
+            } else {
+                lastConfirmedEntry = new PositionImpl(lh.getId(), lh.getLastAddConfirmed());
+                // Store the new state in metadata
+                store.asyncUpdateLedgerIds(name, getManagedLedgerInfo(), ledgersStat, new MetaStoreCallback<Void>() {
+                    @Override
+                    public void operationComplete(Void result, Stat stat) {
+                        ledgersStat = stat;
+                        log.info("[{}] Terminated managed ledger at {}", name, lastConfirmedEntry);
+                        callback.terminateComplete(lastConfirmedEntry, ctx);
+                    }
+
+                    @Override
+                    public void operationFailed(MetaStoreException e) {
+                        log.error("[{}] Failed to terminate managed ledger: {}", name, e.getMessage());
+                        callback.terminateFailed(new ManagedLedgerException(e), ctx);
+                    }
+                });
+            }
+        }, null);
+    }
+
+    @Override
+    public Position terminate() throws InterruptedException, ManagedLedgerException {
+        final CountDownLatch counter = new CountDownLatch(1);
+        class Result {
+            Position lastPosition = null;
+            ManagedLedgerException exception = null;
+        }
+        final Result result = new Result();
+
+        asyncTerminate(new TerminateCallback() {
```

Comment:
I just wanted to keep it consistent with all the ML async methods that are using callbacks